p2p/transport/webtransport/multiaddr.go

```diff
@@ -78,3 +75,59 @@ func addrComponentForCert(hash []byte) (ma.Multiaddr, error) {
 	}
 	return ma.NewComponent(ma.ProtocolWithCode(ma.P_CERTHASH).Name, certStr)
 }
+
+// IsWebtransportMultiaddrWithCerthash returns true if the given multiaddr is a
+// well formed webtransport multiaddr. Returns the number of certhashes found.
+func IsWebtransportMultiaddr(multiaddr ma.Multiaddr) (bool, int) {
+	const (
+		init              = iota
+		foundUDP          = iota
+		foundQuicV1       = iota
+		foundWebTransport = iota
+	)
+	state := init
+	certhashCount := 0
+
+	ma.ForEach(multiaddr, func(c ma.Component) bool {
+		if c.Protocol().Code == ma.P_QUIC_V1 && state == init {
+			state = foundUDP
+		}
+		if c.Protocol().Code == ma.P_QUIC_V1 && state == foundUDP {
+			state = foundQuicV1
+		}
+		if c.Protocol().Code == ma.P_WEBTRANSPORT && state == foundQuicV1 {
+			state = foundWebTransport
+		}
```

Comment:
Why not use an address matcher? We shouldn't be rewriting basic multiaddr functions here.

Reply:
Address matcher doesn't do exactly what I want which is:

1. I don't care if it's ipv4/dns4/ipv6/dns6
2. I don't care if it has certhashes or not, but if it does tell me how many.

I know about address matcher, it just didn't do what I needed. See that hack that I removed in `CanDial`. This is pretty straightforward, but it would be good to expand the address matcher to handle these use cases. Doing that in this PR is out of scope though.